### PR TITLE
Adds conv3d_transpose operation (3D "deconvolution")

### LINKED ISSUE #3012: Conv3d_transpose in TensorFlow
Feature request: Would it be possible to provide a conv3d_transpose, analogous to conv2d_transpose?

### PR DESCRIPTION
**Pull request by daeyun**
This pull request contains the following changes:
- Define gradients for Conv3DBackpropInputV2 and Conv3DBackpropFilterV2.
- Add a new conv3d_transpose operation which uses Conv3DBackpropInputV2.
- Test conv3d_transpose and Conv3DBackpropFilterV2.

Fixes #3012 #150

### CODE REVIEW

**Comment by daeyun** [comment]
I signed the CLA.

File: tensorflow/python/ops/nn_grad.py [hunk 1]
```diff
@@ -82,6 +82,36 @@ def _Conv3DGrad(op, grad):
                                            padding=op.get_attr("padding"))]
 
 
+@ops.RegisterGradient("Conv3DBackpropInputV2")
+def _Conv3DBackpropInputGrad(op, grad):
+  return [None,
+      nn_ops.conv3d_backprop_filter_v2(grad,
```
Comment (vrv) [inline tensorflow/python/ops/nn_grad.py]: weird indentation here, should be aligned with [

File: tensorflow/python/ops/nn_grad.py [hunk 2]
```diff
@@ -82,6 +82,36 @@ def _Conv3DGrad(op, grad):
...
+      nn_ops.conv3d(grad,
+                    op.inputs[1],
+                    strides=op.get_attr("strides"),
+                    padding=op.get_attr("padding"))]
+
+
+@ops.RegisterGradient("Conv3DBackpropFilterV2")
+def _Conv3DBackpropFilterGrad(op, grad):
+  return [
+      nn_ops.conv3d_backprop_input_v2(array_ops.shape(op.inputs[0]),
```
Comment (vrv) [inline tensorflow/python/ops/nn_grad.py]: same here

File: tensorflow/python/ops/nn_ops.py [hunk 3]
```diff
@@ -297,6 +297,68 @@ def conv2d_transpose(value,
                                             name=name)
 
 
+def conv3d_transpose(value,
+                     filter,
+                     output_shape,
+                     strides,
+                     padding="SAME",
+                     name=None):
+  """The transpose of `conv3d`.
```
Comment (vrv) [inline tensorflow/python/ops/nn_ops.py]: Add a comment like with conv2d_transpose that this is sometimes referred to in neural network literature as 'deconvolution'.

**Comment by vrv** [comment]
very nice!  just a few minor comments and then we can test.

**Comment by daeyun** [comment]
Thanks, done.

**Comment by vrv** [comment]
Thanks!
@tensorflow-jenkins test this please

**Comment by KendallWeihe** [comment]
has this been tested and verified?

**Comment by vrv** [comment]
There's a test to validate correctness, yes.  If you think there's a bug in the implementation, please feel free to file an issue.  Thanks!

**Comment by KendallWeihe** [comment]
I'm having a difficult time updating Tensorflow to include this code. I noticed that the function is in source code in the nn_ops.py file, but when I run the pip --upgrade it isn't included. Could I simply copy the function into my existing code, or does it include other dependencies?

**Comment by vrv** [comment]
You could try one of the nightlies that might have this change in it: https://github.com/tensorflow/tensorflow#installation

**Comment by KendallWeihe** [comment]
that had it, thanks!

**Comment by KendallWeihe** [comment]
Can we fix the issue where after a transpose, the dimensions are always <?,?, ?, ?, n_classes> -- can we get rid of the question marks and make them show the actual dimension value?

**Comment by KendallWeihe** [comment]
@vrv it isn't included in the Mac OSX nightly build or the build currently posted to the main install page. Is there any other way I can get it installed on my mac? I ask because of this issue on both of my linux machines *** Error in 'python': free(): invalid pointer outlined in this SO post: http://stackoverflow.com/questions/38129441/tensorflow-conv3d-transpose-error-in-python-free-invalid-pointer

**Comment by vrv** [comment]
http://ci.tensorflow.org/view/Nightly/job/nightly-matrix-cpu/TF_BUILD_CONTAINER_TYPE=CPU,TF_BUILD_IS_OPT=OPT,TF_BUILD_IS_PIP=PIP,TF_BUILD_PYTHON_VERSION=PYTHON2,label=mac-slave/

looks like no mac builds have been created for a while -- pinging @caisq who might know more.

**Comment by KendallWeihe** [comment]
@vrv Do you think this a bug with the latest build, or with the development of `conv3d_transpose` ?

**Comment by vrv** [comment]
I have no idea -- conv3d_transpose is just a wrapper calling an existing function that already likely worked.  I think filing a bug with a reproducible small example would help (small being the clear word here -- it should be a single call to conv3d_transpose that triggers the bug, ideally).

**Comment by KendallWeihe** [comment]
I opened a new issue: https://github.com/tensorflow/tensorflow/issues/3128

I wrote a small network to reproduce the issue -- I hope it was concise enough

**Comment by KendallWeihe** [comment]
@vrv Mac OS nightly build still hasn't been updated -- over a month old

**Comment by vrv** [comment]
I guess there's a stale link somewhere. Try these?

http://ci.tensorflow.org/view/Nightly/job/nightly-matrix-cpu/156/

E.g., http://ci.tensorflow.org/view/Nightly/job/nightly-matrix-cpu/156/TF_BUILD_CONTAINER_TYPE=CPU,TF_BUILD_IS_OPT=OPT,TF_BUILD_IS_PIP=PIP,TF_BUILD_PYTHON_VERSION=PYTHON2,label=mac1-slave/

**Comment by aioie** [comment]
Dear, 
Look  what  I've found on  the web, this is something new I guess, you'll be  excited.  More info here <http://together.mylanpro.com/e4ptuo>

Very truly yours, ziney2y

**Comment by aioie** [comment]
Greetings! 


I've chosen some interesting links  you may  find helpful, here is the full list <http://amazing.phidaux.com/e4nmrpxi>

Regards, ziney2y

File: tensorflow/python/ops/nn_grad.py (hunk 1 above)
Comment (aioie) [inline tensorflow/python/ops/nn_grad.py]: Hi, 


I've been looking  for some useful stuff and found an interesting site with  useful advices, please take a look <http://agree.pumpjuice.com/e4ann>

ziney2y

**Comment by aioie** [comment]
Dear, 

Have you heard about  that new art  gallery opening? I think it's gonna be much  fun,  here  is more info about  the  event <http://motive.stuntwalletproductions.com/e4mxqo>

Thanks for your consideration, ziney2y

**Comment by aioie** [comment]
Hi friend! 

We've  had a nice trip last summer, the pictures are really great.  Just take a look at them here <http://motive.stuntwalletproductions.com/e4zwqavd>


Hugs, ziney2y

**Comment by juanprietob** [comment]
Is this feature implemented for GPU? 
I am getting the following error when tensorflow is compiled with GPU support

`AttributeError: 'module' object has no attribute 'conv3d_transpose'`

**Comment by KendallWeihe** [comment]
@juanprietob I'm not sure this is in production at the moment. I installed from a nightly build.

**Pull request merged**